Data/HashMap/Base.hs

```diff
@@ -488,40 +536,184 @@ bitmapIndexedOrFull b ary
+
+-- Insert optimized for the case when we know the key is not in the map.
+--
+-- It is only valid to call this when the key does not exist in the map.
+--
+-- We can skip:
+--  - the key equality check on a Leaf
+--  - check for its existance in the array for a hash collision
+insertNewKey :: Hashable k
+             => Hash -> k -> v -> Int -> HashMap k v -> HashMap k v
+insertNewKey !h !k x !_ Empty = Leaf h (L k x)
+insertNewKey h k x s (Leaf hy l@(L ky y))
+    | hy == h = collision h l (L k x)
+    | otherwise = runST (two s h k x hy ky y)
+insertNewKey h k x s t@(BitmapIndexed b ary)
+    | b .&. m == 0 =
+        let !ary' = A.insert ary i $! Leaf h (L k x)
+        in bitmapIndexedOrFull (b .|. m) ary'
+    | otherwise =
+        let !st  = A.index ary i
+            !st' = insertNewKey h k x (s+bitsPerSubkey) st
+        in if st' `ptrEq` st
+           then t
+           else BitmapIndexed b (A.update ary i st')
+  where m = mask h s
+        i = sparseIndex b m
+insertNewKey h k x s t@(Full ary) =
+    let !st  = A.index ary i
+        !st' = insertNewKey h k x (s+bitsPerSubkey) st
+    in if st' `ptrEq` st
+        then t
+        else Full (update16 ary i st')
+  where i = index h s
+insertNewKey h k x s t@(Collision hy v)
+    | h == hy   = Collision h (snocNewLeaf (L k x) v)
+    | otherwise =
+        insertNewKey h k x s $ BitmapIndexed (mask hy s) (A.singleton t)
   where
-    h0 = hash k0
-    go !h !k x !_ Empty = Leaf h (L k x)
-    go h k x s t@(Leaf hy l@(L ky y))
-        | hy == h = if ky == k
-                    then if x `ptrEq` y
-                         then t
-                         else Leaf h (L k x)
-                    else collision h l (L k x)
-        | otherwise = runST (two s h k x hy ky y)
-    go h k x s t@(BitmapIndexed b ary)
-        | b .&. m == 0 =
-            let !ary' = A.insert ary i $! Leaf h (L k x)
-            in bitmapIndexedOrFull (b .|. m) ary'
-        | otherwise =
-            let !st  = A.index ary i
-                !st' = go h k x (s+bitsPerSubkey) st
-            in if st' `ptrEq` st
-               then t
-               else BitmapIndexed b (A.update ary i st')
-      where m = mask h s
-            i = sparseIndex b m
-    go h k x s t@(Full ary) =
+    snocNewLeaf :: Leaf k v -> A.Array (Leaf k v) -> A.Array (Leaf k v)
+    snocNewLeaf leaf ary = A.run $ do
+      let n = A.length ary
+      mary <- A.new_ (n + 1)
+      A.copy ary 0 mary 0 n
+      A.write mary n leaf
+      return mary
+{-# INLINABLE insertNewKey #-}
+
+
+-- Insert optimized for the case when we know the key is in the map and does not
+-- have a hash collision.
+--
+-- It is only valid to call this when the key exists in the map and does not
+-- have a hash collision with another key. This information can be obtained
+-- from 'lookupRecordCollision'.
+--
+-- We can skip:
+--  - the key equality check on a Leaf, we know the leaf must be for this key.
+insertExistingKeyNoCollision
```

Comment:
I think you can merge this with the version for collisions. Just use a negative index when there's not supposed to be a collision; you can test the sign when you get to the bottom for sanity checking.